### PR TITLE
fix: Fix and refactor Date

### PR DESCRIPTION
**Pull request by MaxGraey**
- Some post improvements and refactors for #1768
- Add getUTCDay

- [x] I've read the contributing guidelines

### CODE REVIEW

File: std/assembly/date.ts [hunk 1]
```diff
@@ -1,6 +1,20 @@
...
+@inline const
+  MILLIS_PER_DAY    = 1000 * 60 * 60 * 24,
+  MILLIS_PER_HOUR   = 1000 * 60 * 60,
+  MILLIS_PER_MINUTE = 1000 * 60,
+  MILLIS_PER_SECOND = 1000;
+
+// ymdFromEpochDays returns values via globals to avoid allocations
+// @ts-ignore: decorator
+@lazy let
+  year: i32,
```
Comment (dcodeIO) [inline std/assembly/date.ts:14]: Can perhaps reduce by one global here, by returning say `year` from the function

**Comment by MaxGraey** [comment]
@ColinEberhardt It seems `ymdFromEpochDays` have some edge case for negative values like:
```ts
ymdFromEpochDays(-8640000000000000 + 1);
```
result should be:
```
day: 20
hours: 0
minutes: 0
seconds: 0
millis: 1
```
actual:
```
day: 21
hours: -23
minutes: -59
seconds: -59
millis: -999
```

File: std/assembly/date.ts [hunk 2]
```diff
@@ -1,7 +1,22 @@
-import { E_VALUEOUTOFRANGE } from "util/error";
+import { E_INVALIDDATE } from "util/error";
 import { now as Date_now } from "./bindings/Date";
 
+// @ts-ignore: decorator
+@inline const
```
Comment (SebastianSpeitel) [inline std/assembly/date.ts:5]: Should we make these @lazy as well? Or is that implied by @inline?

Comment (MaxGraey) [reply std/assembly/date.ts:5]: No, const globals usually not needed lazy due to it will always removed by binaryen

File: std/assembly/date.ts [hunk 3]
```diff
@@ -11,155 +26,173 @@ export class Date {
...
-        second: i32 = 0,
-        millisecond: i32 = 0;
-    let dateString: string;
+  // It can parse only ISO 8601 inputs: YYYY-MM-DDThh:mm-ss.nnnnZ
+  @inline static parse(dateString: string): Date {
+    return this.fromString(dateString);
+  }
 
-    if (dateTimeString.includes("T")) {
+  static fromString(dateTimeString: string): Date {
```
Comment (SebastianSpeitel) [inline std/assembly/date.ts:44]: Should this be private, since it's not part of the Date spec?

Comment (MaxGraey) [reply std/assembly/date.ts:44]: Yeah, it's not part of Date. But currently we can't use ad-hoc polymorphism for constructors so it's workaround solution for `new Date("2021-04-15")`.

**Review by torch2424 — APPROVED** [review]
I did a quick review, and this PR looks good to me :smile: 

Also, shoutout to @ColinEberhardt for kicking this off! :smile: :tada:

**Review by ColinEberhardt — APPROVED** [review]
Looks spot on to me - thanks @MaxGraey

**Comment by MaxGraey** [comment]
@ColinEberhardt thanks! I'll merge this when node nightly build (canary) will be available

**Pull request merged**